### PR TITLE
add test studio

### PR DESCRIPTION
**Pull request by Aratramba**
added a bunch of tests and a GitHub action #8 

Also made some interface tweaks:

- made edit dialog flexible vertically
- added empty screen message
- added aria labels to buttons

image

test

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -40,10 +42,25 @@
...
     "@babel/preset-react": "7.12.10",
     "@babel/preset-typescript": "7.12.7",
+    "@sanity/base": "^2.3.2",
+    "@sanity/cli": "^2.2.6",
+    "@sanity/components": "^2.2.6",
+    "@sanity/core": "^2.2.6",
+    "@sanity/default-layout": "^2.3.2",
+    "@sanity/default-login": "^2.2.6",
+    "@sanity/desk-tool": "^2.3.2",
+    "@sanity/vision": "^2.2.6",
```
Comment (dennispassway) [inline package.json:52]: Do we use vision?

File: package.json [hunk 2]
```diff
@@ -40,10 +42,25 @@
...
     "@types/react": "17.0.0",
     "@types/styled-components": "^5.1.7",
     "babel-plugin-styled-components": "^1.12.0",
+    "dotenv": "^8.2.0",
+    "jest-playwright-preset": "^1.4.5",
+    "jest": "^26.6.3",
+    "playwright": "^1.8.0",
     "prettier": "^2.2.1",
+    "prop-types": "^15.7",
```
Comment (dennispassway) [inline package.json:61]: do we use prop-types?

File: package.json [hunk 3]
```diff
@@ -40,10 +42,25 @@
...
+    "react-dom": "^17.0",
+    "react": "^17.0",
```
Comment (dennispassway) [inline package.json:63]: react can be removed, since it is already listes as peer dependency

File: package.json [hunk 4]
```diff
@@ -40,10 +42,25 @@
     "@babel/preset-env": "^7.12.11",
     "@babel/preset-react": "7.12.10",
     "@babel/preset-typescript": "7.12.7",
+    "@sanity/base": "^2.3.2",
```
Comment (dennispassway) [inline package.json:45]: already listed as peer dependecny

File: src/components/DeleteModal.tsx [hunk 5]
```diff
@@ -59,7 +59,9 @@ export const DeleteModal = ({ assets, loading, handleError, onClose, onDeleteCom
   return (
     <Modal onClose={onClose}>
       <StyledContainer>
-        <StyledTitle>Are you sure you want to delete {plural ? 'these assets' : 'this asset'}?</StyledTitle>
+        <StyledTitle>
+          Are you sure you want to delete {plural ? `these ${assets.length} assets` : 'this asset'}?
```
Comment (dennispassway) [inline src/components/DeleteModal.tsx:63]: 'these' could now be removed. nice improvement 🌱

File: src/components/MediaLibrary.tsx [hunk 6]
```diff
@@ -121,14 +125,18 @@ export const MediaLibrary = ({
...
-          <ViewElement
-            assets={assets}
-            onDoubleClick={onDoubleClick}
-            onDragStart={onDragStart}
-            onMediaItemClick={onMediaItemClick}
-            selectedAssets={selectedAssets}
-            setIsDraggingMediaItem={setIsDraggingMediaItem}
-          />
+          {!Boolean(assets.length) && !loading ? (
+            <StyledMessage>No files yet</StyledMessage>
```
Comment (dennispassway) [inline src/components/MediaLibrary.tsx:129]:
```suggestion
            <StyledMessage>No assets yet</StyledMessage>
```

File: studio/package.json [hunk 7]
```diff
@@ -0,0 +1,9 @@
+{
```
Comment (dennispassway) [inline studio/package.json:1]: Does this belogn here?

File: studio/plugins/.gitkeep [hunk 8]
```diff
@@ -0,0 +1 @@
+User-specific packages can be placed here
```
Comment (dennispassway) [inline studio/plugins/.gitkeep:1]: This file can be deleted.

File: studio/sanity.json [hunk 9]
```diff
@@ -0,0 +1,31 @@
+{
+  "root": true,
+  "api": {
+    "projectId": "q74m7lt8",
```
Comment (dennispassway) [inline studio/sanity.json:4]:
```suggestion
    "projectId": null,
```

File: studio/sanity.json [hunk 10]
```diff
@@ -0,0 +1,31 @@
...
+    "dataset": "production"
```
Comment (dennispassway) [inline studio/sanity.json:5]:
```suggestion
    "dataset": null
```

File: studio/sanity.json [hunk 11]
```diff
@@ -0,0 +1,31 @@
...
+  },
+  "plugins": [
+    "@sanity/base",
+    "@sanity/components",
+    "@sanity/default-layout",
+    "@sanity/default-login",
+    "@sanity/desk-tool",
+    "media-library"
+  ],
+  "env": {
```
Comment (dennispassway) [inline studio/sanity.json:18]: vision can be removed

File: test/asset-source.test.js [hunk 12]
```diff
@@ -0,0 +1,228 @@
+require('regenerator-runtime/runtime');
+const path = require('path');
+
+import { formatDate, formatSize } from '../src/shared/utils';
+
+// Needs to be higher than the default Playwright timeout
```
Comment (dennispassway) [inline test/asset-source.test.js:6]: add why

File: test/asset-source.test.js [hunk 13]
```diff
@@ -0,0 +1,228 @@
...
+jest.setTimeout(40 * 1000);
```
Comment (dennispassway) [inline test/asset-source.test.js:7]:
```suggestion
jest.setTimeout(40000);
```

File: test/asset-source.test.js [hunk 14]
```diff
@@ -0,0 +1,228 @@
...
+    await expect(page).toHaveText('h2', 'Filters');
+  });
+
+  it('should upload images', async () => {
+    await page.setInputFiles(
+      'input[type="file"]',
+      IMAGES.map((img) => path.join(__dirname, 'fixtures', img)),
+      {}
+    );
+    await page.waitForTimeout(2000);
```
Comment (dennispassway) [inline test/asset-source.test.js:47]: This fails on slow internet speeds right?

File: test/asset-source.test.js [hunk 15]
```diff
@@ -0,0 +1,228 @@
...
+    await page.click('text=Save changes');
+    await page.fill('[placeholder="Search by filename, title, alt or tag"]', '');
+    await expect(page).toHaveSelectorCount('[draggable]', 4);
+    await page.click('text=TEST_TAG1');
+    await expect(page).toHaveSelectorCount('[draggable]', 1);
+    await page.click('text=Clear all filters');
+    await expect(page).toHaveSelectorCount('[draggable]', 4);
+  });
+
+  // it('should remove unused tag', async () => {
```
Comment (dennispassway) [inline test/asset-source.test.js:120]: This should be removed or fixed before merging

File: test/asset-source.test.js [hunk 16]
```diff
@@ -0,0 +1,228 @@
...
+    await expect(page).toHaveText('[role="dialog"]', 'Are you sure you want to delete this asset?');
+    await page.click('[role="dialog"] button');
+    await page.waitForTimeout(2000);
+
+    await page.click(`[draggable]`, { modifiers: ['Shift'] });
+    await page.click(`[draggable]:nth-of-type(2)`, { modifiers: ['Shift'] });
+    await page.click('text=Delete Asset');
+    await expect(page).toHaveText('[role="dialog"]', 'Are you sure you want to delete these 2 assets?');
+    await page.click('[role="dialog"] button');
+    await page.waitForTimeout(2000);
```
Comment (dennispassway) [inline test/asset-source.test.js:214]: Same for all waitForTimeout, maybe put the time in a variable up top with a comment that this is the max time the requests may take, so slower internet conneciton can simply up the variable

**Review by dennispassway — CHANGES_REQUESTED** [review]
Overall really nice. I think th estudio should not be mixed with the plugin tho and it should be in a seperate test fodler

File: package.json (hunk 1 above)
Comment (Aratramba) [reply package.json:52]: pretty handy for development

File: studio/package.json (hunk 7 above)
Comment (Aratramba) [reply studio/package.json:1]: there needs to be a package.json inside a studio folder

File: studio/sanity.json (hunk 9 above)
Comment (Aratramba) [reply studio/sanity.json:4]: I think we can leave these here so you can quickstart development. If you need to change it, you can use sanity reserved variables to override them https://www.sanity.io/docs/studio-environment-variables#reserved-variable-names-cc239c4841f0

File: README.md [hunk 17]
```diff
@@ -199,3 +199,16 @@ Example with asset fields listed and custom fields added:
...
 To contribute a theme, add it in `themes/[themename].ts`.
 If you run into problems or have feature requests, please create an issue or pull request and I'll look into it as soon as I can.
+
+## Development
+To test the studio while developing, make a symlink inside the /studio/plugins that links to the root directory. Then run `sanity start` in the studio folder. 
+
+Run `yarn dev` in the root folder to make changes.
+
+## Testing
+Create a Sanity bot token with write access at manage.sanity.io. Add a dataset named 'testing'. Then make a .env file  inside the root folder:
```
[the remark this answers is not in the document]
Comment (Aratramba) [reply README.md]: these can be overridden by env vars https://www.sanity.io/docs/studio-environment-variables#reserved-variable-names-cc239c4841f0

File: .gitignore [hunk 18]
```diff
@@ -2,4 +2,6 @@
 *.log*
 lib
 logs
-node_modules
\ No newline at end of file
+node_modules
+test/studio/plugins/sanity-plugin-media-library
+.env
```
Comment (dennispassway) [inline .gitignore:7]:
```suggestion
.env*
```

to prevent all sort of env files from being commited

File: README.md [hunk 19]
```diff
@@ -199,3 +199,17 @@ Example with asset fields listed and custom fields added:
 ## Contributing
 To contribute a theme, add it in `themes/[themename].ts`.
 If you run into problems or have feature requests, please create an issue or pull request and I'll look into it as soon as I can.
+
+## Development
+Run `yarn dev` in the root folder to make changes.
+
+To test the studio while developing, make a symlink inside the `/test/studio/plugins` folder that links to the root directory. Then run `sanity start` in `test/studio`.
```
Comment (dennispassway) [inline README.md:206]:
```suggestion
Optional: To use the included studio while developing, make a symlink inside the `/test/studio/plugins` folder that links to the root directory. Then run `sanity start` in `test/studio`. You can also add the plugin to the plugins directory in a Sanity studio of your own.
```

File: README.md [hunk 20]
```diff
@@ -199,3 +199,17 @@ Example with asset fields listed and custom fields added:
...
 If you run into problems or have feature requests, please create an issue or pull request and I'll look into it as soon as I can.
+
+## Development
+Run `yarn dev` in the root folder to make changes.
+
+To test the studio while developing, make a symlink inside the `/test/studio/plugins` folder that links to the root directory. Then run `sanity start` in `test/studio`. 
+
+
+## Testing
+Create a Sanity bot token with write access at [manage.sanity.io](manage.sanity.io). Add a dataset named 'testing'. Then make a .env file  inside the `/test` folder:
```
Comment (dennispassway) [inline README.md:210]: The dataset testing should also be in the env file?

File: test/studio/package.json [hunk 21]
```diff
@@ -0,0 +1,8 @@
+{
+  "name": "medialibrary-studio",
+  "private": true,
+  "version": "1.0.0",
+  "main": "package.json",
+  "author": "Arjen <aratramba@gmail.com>",
```
Comment (dennispassway) [inline test/studio/package.json:6]: Sure you want your email to appear here?

File: test/studio/sanity.json [hunk 22]
```diff
@@ -0,0 +1,31 @@
+{
+  "root": true,
+  "api": {
+    "projectId": "q74m7lt8",
```
Comment (dennispassway) [inline test/studio/sanity.json:4]: these should be null by default i think

File: test/studio/sanity.json [hunk 23]
```diff
@@ -0,0 +1,31 @@
...
+    "@sanity/components",
+    "@sanity/default-layout",
+    "@sanity/default-login",
+    "@sanity/desk-tool",
+    "media-library"
+  ],
+  "env": {
+    "development": {
+      "plugins": [
+        "@sanity/vision"
```
Comment (dennispassway) [inline test/studio/sanity.json:21]: cant we remove vision?

**Review by dennispassway — CHANGES_REQUESTED** [review]

**Pull request merged**